pysnark/fixedpoint.py

```diff
@@ -28,107 +28,171 @@
+        """
+        Converts an integer, float, or LinComb into the internal fixed point representation
+        and converts the input into a LinCombFxp
+        """
+        if isinstance(val, LinCombFxp):
             return val
-        elif isinstance(val,LinComb):
-            return cls.fromvar(val, True)
-        else:
-            v = _tofxpval(val, True)
-            return cls(LinComb.ONE*v)
+        if isinstance(val, LinComb):
+            val = LinCombFxp.add_scaling(val)
+            return LinCombFxp(val)
+        if isinstance(val, int) or isinstance(val, float):
+            val = LinCombFxp.add_scaling(val)
+            return LinCombFxp(ConstVal(val))
+        raise RuntimeError("Wrong type for _ensurefxp")
     
     def __add__(self, other):
-        other = self._ensurefxp(other)
-        return LinCombFxp(self.lc+other.lc)
-        
+        """
+        Adds a LinCombFxp with an integer, float, LinComb, or another LinCombFxp
+        Costs 0 constraints
+        """
+        if isinstance(other, int) or isinstance(other, float) or isinstance(other, LinComb):
+            other = LinCombFxp.add_scaling(other)
+            return LinCombFxp(self.lc + other)
+        if isinstance(other, LinCombFxp):
+            return LinCombFxp(self.lc + other.lc)
+        return NotImplemented
+
     __radd__ = __add__
     
     def __sub__(self, other):
-        return self+(-other)
+        return self + (-other)
     
     def __rsub__(self, other):
-        return other+(-self)    
+        return other + (-self)
 
     def __mul__(self, other):
-        if is_base_value(other):
-            return LinCombFxp(self.lc*other)
-        
-        other = self._ensurefxp(other)
-        if other is NotImplemented: return NotImplemented
-        
-        ret = LinCombFxp(PrivVal((self.lc.value*other.lc.value+(1<<(LinCombFxp.res-1)))>>LinCombFxp.res))
-            
-        # mul error: should be in [-2^f,2^f]
-        diff = (1<<LinCombFxp.res)*ret.lc-self.lc*other.lc
-        (diff + (1<<LinCombFxp.res)).assert_positive(LinCombFxp.res+1)
-        ((1<<LinCombFxp.res) - diff).assert_positive(LinCombFxp.res+1)
-
-        return ret        
-    
+        """
+        Multiplies a LinCombFxp with an integer, float, LinComb, or another LinCombFxp
+        Costs 0 constraints to multiply with an integer or a float
+        Costs 1 constraint to multiply with a LinComb or LinCombFxp
+        """
+        if isinstance(other, int):
+            return LinCombFxp(self.lc * other)
+        if isinstance(other, float):
+            other = LinCombFxp.add_scaling(other)
+            return LinCombFxp((self.lc * other) / (1 << resolution))
+        if isinstance(other, LinComb):
+            other = LinCombFxp.add_scaling(other)
+            return LinCombFxp((self.lc * other) / (1 << resolution))
+        if isinstance(other, LinCombFxp):
+            return LinCombFxp((self.lc * other.lc) / (1 << resolution))
+        return NotImplemented

```

Comment:
This should work. It will const bit_length constraints due to the right-shift (at least I assume (1<<resolution) is implemented with a right shift in runtime.py? otherwise it is more costly). The previous code did two assert_positive's of size equal to the resolution. So it seems that depending the resolution and the bit length either of the two could be more efficient.. There would also be differences in the sizes of the inputs that they would accept.. I would have to think a bit more about which version would be preferable in which cases.

Reply:
Good point. `LinComb.__rshift__` uses a division right now instead of `to_bits`. I'll have to change that and change the fixed point code to use `>> (1 << resolution)`.

Could you explain the constraints you had before? Are they still required here?
```
diff = (1<<LinCombFxp.res)*ret.lc-self.lc*other.lc
(diff + (1<<LinCombFxp.res)).assert_positive(LinCombFxp.res+1)
((1<<LinCombFxp.res) - diff).assert_positive(LinCombFxp.res+1)
```

Reply:
Actually, I did some digging and it turns out that `LinCombFxp.__mul__` costs only 1 constraint, even with the division. I traced it to `LinComb.__truediv__` costing 0 constraints to divide by a public integer, and `(1 << resolution)` being a public integer.

Could you check that `LinComb.__truediv__` is correct?

Reply:
LinComb.__truediv__ as it is now implements proper division, meaning it only works if the division does not have a remainder. So you can divide 8 by 4 (and so do a right-shift by two), but not 9 because this division has remainder one. So you cannot use __truediv__ to implement a right-shift, it should be __floordiv__. I think for a right-shift there is really no other way than doing a bit decomposition one way or the other

Reply:
Is `floordiv` really necessary? For example, if we multiply two fixedpoints, then we have `res = (a * s) * (b * s)` where `s` is the scaling `1 << resolution`. Then `s` always divides `res` evenly so we can just use `truediv`, right?

I believe the same argument works for float * fixedpoint.

Reply:
I see. 
I'll implement it using floordiv for now. Using a direct rshift with bit decomposition is cheaper, but we easily run into trouble because of to_bits since the numbers are often larger than `2 ** bitlength`.